app/code/Magento/Sales/Model/ShipOrder.php

```diff
@@ -149,7 +149,8 @@ public function execute(
         ShipmentCreationArgumentsInterface $arguments = null
     ) {
         $connection = $this->resourceConnection->getConnection('sales');
-        $order = $this->orderRepository->get($orderId);
+        $connection->beginTransaction();
+        $order = $this->orderRepository->getForUpdate($orderId);
```

Comment:
In case if the service call will fail due to WAIT LOCK timeout the service has to communicate an explicit message that the operation retryable.

Reply:
The query will block until the lock has been released when a concurrent transaction has been completed. If it does encounter a "Wait Lock Timeout Exceeded" error from MariaDB an exception will be thrown with an appropriate message.

Reply:
Is there an example of processing such errors in other places in the system?

Reply:
The service should not return exception from the database tier. My point was that you need to catch it, wrap into exception proper to the layer, and rethrow.

Reply:
It would make sense to do that exception type conversion in the repository, wouldn't it?

Reply:
I agree @akaplya. This isn't practiced throughout the system but as we're developing we should go ahead and implement better practices.

Reply:
Implement error handling in the OrderRepository similar to what is found in the ProductRepository (https://github.com/magento/magento2/blob/89d3cda99975c0db2967165311184be84d26e9b4/app/code/Magento/Catalog/Model/ProductRepository.php#L832-L837).